File: Алгоритмы и структуры данных/Краскал на Java/UniqueArray.java
package org.example;

public class UniqueArray<K, V>
{
    private class Node<K, V>
    {
        K key;
        V value;

        Node(K key, V value)
        {
            this.key = key;
            this.value = value;
        }
    }

    DynamicArray<Node> array = new DynamicArray<Node>();

    public boolean isEmpty()
    {
        return array.size() == 0;
    }

    public boolean containsKey(K key)
    {
        for (int i = 0; i < array.size(); i++)
        {
            if (array.get(i).key == key)
            {
                return true;
            }
        }

        return false;
    }

    public boolean containsValue(V value)
    {
        for (int i = 0; i < array.size(); i++)
        {
            if (array.get(i).value == value)
            {
                return true;
            }
        }

        return false;
    }

    public boolean put(K key, V value)
    {
        for (int i = 0; i < array.size(); i++)
        {
            if (key == array.get(i).key)
            {
                array.set(new Node<>(key, value), i);
                return true;
            }
        }

        array.add(new Node<>(key, value));
        return true;
    }

    public void remove(K key)
    {
        for (int i = 0; i < array.size(); i++)
        {
            if (array.get(i).key == key)
            {
                array.remove(i);
                break;
            }
        }
    }

    public void remove(K key, V value)
    {
        for (int i = 0; i < array.size(); i++)
        {
            if (array.get(i).key == key && array.get(i).value == value)
            {
                array.remove(i);
                break;
            }
        }
    }

    public void clear()
    {
        array.clear();
    }

    public int size()
    {
        return array.size();
    }

    public V get(K key)
    {
        for (int i = 0; i < array.size(); i++)
        {
            if (array.get(i).key == key)
            {
                return (V) array.get(i).value;
            }
        }

        return null;
    }
}

File: Алгоритмы и структуры данных/Краскал/unique_array.cpp
#include "dynamic_array.cpp"

template<typename K, typename V>
class unique_array {
private:
    struct node {
        K key;
        V value;

        node() {

        }

        node(K key, V value) {
            this->key = key;
            this->value = value;
        }
    };

    dynamic_array<node> array;
    node none;
public:
    bool is_empty() {
        return array.get_size();
    }

    bool put(K key, V value) {
        for (int i = 0; i < array.get_size(); i++) {
            if (key == array.get(i).key) {
                node tmp(key, value);
                array.set(tmp, i);
                return true;
            }
        }

        node tmp(key, value);
        array.add(tmp);
        return true;
    }

    void remove(K key) {
        for (int i = 0; i < array.get_size(); i++) {
            if (array.get(i).key == key) {
                array.remove(i);
                break;
            }
        }
    }


    int get_size() {
        return array.get_size();
    }

    V get(K key) {
        for (int i = 0; i < array.get_size(); i++) {
            if (array.get(i).key == key) {
                return array.get(i).value;
            }
        }

        return none.value;
    }
};